lib/smooth/core/io/spi/Master.cpp

```diff
@@ -23,63 +23,101 @@ namespace smooth::core::io::spi
 {
     static constexpr const char* log_tag = "SPIMaster";
 
-    Master::Master(
-        spi_host_device_t host,
-        SPI_DMA_Channel dma_channel,
-        gpio_num_t mosi,
-        gpio_num_t miso,
-        gpio_num_t clock,
-        int transfer_size,
-        gpio_num_t quadwp_io_num,
-        gpio_num_t quadhd_io_num
-        )
-            : host(host),
-              dma_channel(dma_channel)
+    // Declare static variables
+    bool Master::spi_bus_initialized = false;
+    bool Master::hspi_initialized = false;
+    bool Master::vspi_initialized = false;
+    std::mutex Master::guard{};
+    spi_bus_config_t Master::bus_config{};
+    spi_host_device_t Master::spi_host;
+    SPI_DMA_Channel Master::dma_channel;
+
+    bool Master::initialize(spi_host_device_t host,
+                            SPI_DMA_Channel dma_chl,
+                            gpio_num_t mosi,
+                            gpio_num_t miso,
+                            gpio_num_t clock,
+                            int transfer_size,
+                            gpio_num_t quadwp_io_num,
+                            gpio_num_t quadhd_io_num)
     {
+        std::lock_guard<std::mutex> lock(guard);
```

Comment:
When we have the lock, shouldn't we check if we're already initialized before doing anything?

Reply:
I think now that we are keep count of the number of vspi_initialized and hspi_initialized we can't just check if either are initialized.

Reply:
this is what Master.cpp looks like now - lots of duplicate code.
```
#include "smooth/core/io/spi/Master.h"
#include "smooth/core/logging/log.h"

using namespace smooth::core::logging;

namespace smooth::core::io::spi
{
    static constexpr const char* log_tag = "SPIMaster";

    // Declare static variables
    bool Master::hspi_initialized = false;
    bool Master::vspi_initialized = false;
    std::mutex Master::guard{};
    spi_bus_config_t Master::bus_config{};
    spi_host_device_t Master::spi_host;
    SPI_DMA_Channel Master::dma_channel;
    uint8_t Master::hspi_initialized_count = 0;
    uint8_t Master::vspi_initialized_count = 0;

    bool Master::initialize(spi_host_device_t host,
                            SPI_DMA_Channel dma_chl,
                            gpio_num_t mosi,
                            gpio_num_t miso,
                            gpio_num_t clock,
                            int transfer_size,
                            gpio_num_t quadwp_io_num,
                            gpio_num_t quadhd_io_num)
    {
        std::lock_guard<std::mutex> lock(guard);

        spi_host = host;
        dma_channel = dma_chl;

        bus_config.mosi_io_num = mosi;
        bus_config.miso_io_num = miso;
        bus_config.sclk_io_num = clock;
        bus_config.quadwp_io_num = quadwp_io_num;
        bus_config.quadhd_io_num = quadhd_io_num;
        bus_config.max_transfer_sz = transfer_size;

        bool initialized = false;

        if (spi_host == VSPI_HOST)
        {
            initialized = do_vspi_initialization();
        }

        if (spi_host == HSPI_HOST)
        {
            initialized = do_hspi_initialization();
        }

        return initialized;
    }

    bool Master::do_vspi_initialization()
    {
        if (!vspi_initialized)
        {
            vspi_initialized = spi_bus_initialize(spi_host, &bus_config, dma_channel) == ESP_OK;

            if (!vspi_initialized)
            {
                Log::error(log_tag, "VSPI_HOST Initialization failed");
            }
            else
            {
                Log::info(log_tag, "VSPI initialized, Host {}, DMA {}", spi_host, dma_channel);
            }
        }
        
        // increment the vspi_initialized_count so we know how many spi-devices are on the vspi-bus
        if (vspi_initialized)
        {
            vspi_initialized_count++;
        }

        return vspi_initialized;
    }

    bool Master::do_hspi_initialization()
    {
        if (!hspi_initialized)
        {
            hspi_initialized = spi_bus_initialize(spi_host, &bus_config, dma_channel) == ESP_OK;

            if (!hspi_initialized)
            {
                Log::error(log_tag, "HSPI_HOST Initialization failed");
            }
            else
            {
                Log::info(log_tag, "HSPI initialized, Host {}, DMA {}", spi_host, dma_channel);
            }
        }

        // increment the hspi_initialized_count so we know how many spi-devices are on the hspi-bus
        if (hspi_initialized)
        {
            hspi_initialized_count++;
        }

        return hspi_initialized;
    }

    void Master::deinitialize(spi_host_device_t spi_host)
    {
        std::lock_guard<std::mutex> lock(guard);

        if ((spi_host == VSPI_HOST) & vspi_initialized)
        {
            vspi_initialized_count--;

            if (vspi_initialized_count == 0)
            {
                vspi_initialized = false;
                spi_bus_free(VSPI_HOST);
            }
        } 
       

        if ((spi_host == HSPI_HOST) & hspi_initialized)
        {
            hspi_initialized_count--;

            if (hspi_initialized_count == 0)
            {
                hspi_initialized = false;
                spi_bus_free(VSPI_HOST);
            }
        } 
    }
}
```

Reply:
Will supply updated code shortlu